internal/k8s_watcher/watcher.go

```diff
@@ -98,26 +94,28 @@ func WithBlocktxTicker(t Ticker) func(*Watcher) {
 }
 
 func (c *Watcher) Start() error {
-	c.watchMetamorph()
+	ctx, cancel := context.WithCancel(context.Background())
+	c.shutdownMetamorph = cancel
+	c.WaitGroup.Add(1)
+	c.watchMetamorph(ctx)
 
-	c.watchBlocktx()
+	ctx, cancel = context.WithCancel(context.Background())
+	c.shutdownBlocktx = cancel
+	c.WaitGroup.Add(1)
+	c.watchBlocktx(ctx)
 
 	return nil
 }
 
-func (c *Watcher) watchBlocktx() {
+func (c *Watcher) watchBlocktx(ctx context.Context) {
 	go func() {
-		defer func() {
-			if r := recover(); r != nil {
-				c.logger.Error("Recovered from panic", "panic", r, slog.String("stacktrace", string(debug.Stack())))
-			}
-			c.shutdownBlocktxComplete <- struct{}{}
-		}()
-
 		var runningPods map[string]struct{}
 
 		for {
 			select {
+			case <-ctx.Done():
+				c.WaitGroup.Done()
```

Comment:
Please use `defer` for the call to `c.WaitGroup.Done()`

Reply:
And move it outside the loop

Reply:
Fixed.